### PR TITLE
Actions and triggers convertkit

### PR DESCRIPTION
**Pull request by bush4u**

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/convertkit/actions/add-subscriber-to-form/add-subscriber-to-form.mjs [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import convertkit from "../../convertkit.app.mjs";
+
+export default {
+  key: "convertkit-add-subscriber-to-form",
+  name: "Add subscriber to a form",
+  description: "Add subscriber to a form. [See docs here](https://developers.convertkit.com/#add-subscriber-to-a-form)",
+  version: "0.0.2",
```
Comment (lcaresia) [inline components/convertkit/actions/add-subscriber-to-form/add-subscriber-to-form.mjs:7]:
```suggestion
  version: "0.0.1",
```

File: components/convertkit/actions/add-subscriber-to-form/add-subscriber-to-form.mjs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    form: {
+      propDefinition: [
+        convertkit,
+        "form",
+      ],
+    },
+  },
+  async run({ $ }) {
+    const response = await this.convertkit.addSubscriberToForm(this.subscriber, this.form, $);
+    response && $.export("$summary", "Successfully added subscriber to form");
```
Comment (lcaresia) [inline components/convertkit/actions/add-subscriber-to-form/add-subscriber-to-form.mjs:29]: Use a normal `if` here.

```suggestion
    response && $.export("$summary", "Successfully added subscriber to form");
```

File: components/convertkit/actions/list-subscribers/list-subscribers.mjs [hunk 3]
```diff
@@ -0,0 +1,67 @@
+import convertkit from "../../convertkit.app.mjs";
+import constants from "../../common/constants.mjs";
+
+export default {
+  key: "convertkit-list-subscribers",
+  name: "List subscribers",
+  description: "Returns a list of your subscribers. [See docs here](https://developers.convertkit.com/#list-subscribers)",
+  version: "0.0.2",
```
Comment (lcaresia) [inline components/convertkit/actions/list-subscribers/list-subscribers.mjs:8]:
```suggestion
  version: "0.0.1",
```

File: components/convertkit/actions/list-subscribers/list-subscribers.mjs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    const data = [];
+    const paginator = this.convertkit.paginate({
+      $,
+      fn: this.convertkit.listSubscribers,
+      payload,
+    }, "subscribers");
+    for await (const item of paginator) {
+      data.push(item);
+    }
+    console.log(data);
```
Comment (lcaresia) [inline components/convertkit/actions/list-subscribers/list-subscribers.mjs:63]:
```suggestion
```

File: components/convertkit/actions/list-subscribers/list-subscribers.mjs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    data.length && $.export("$summary", "Successfully listed subscribers");
```
Comment (lcaresia) [inline components/convertkit/actions/list-subscribers/list-subscribers.mjs:64]: Use a normal `if` here.

File: components/convertkit/actions/view-single-subscribers/view-single-subscribers.mjs [hunk 6]
```diff
@@ -0,0 +1,26 @@
+import convertkit from "../../convertkit.app.mjs";
+
+export default {
+  key: "convertkit-view-single-subscribers",
+  name: "View a Single Subscribers",
+  description: "Returns data for a single subscriber. [See docs here](https://developers.convertkit.com/#view-a-single-subscriber)",
+  version: "0.0.2",
```
Comment (lcaresia) [inline components/convertkit/actions/view-single-subscribers/view-single-subscribers.mjs:7]:
```suggestion
  version: "0.0.1",
```

File: components/convertkit/actions/view-single-subscribers/view-single-subscribers.mjs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        "subscriber",
+        () => ({
+          returnField: "id",
+        }),
+      ],
+    },
+  },
+  async run({ $ }) {
+    const response = await this.convertkit.getSubscriber(this.subscriber, $);
+    response && $.export("$summary", "Successfully found subscriber");
```
Comment (lcaresia) [inline components/convertkit/actions/view-single-subscribers/view-single-subscribers.mjs:23]: Use a normal `if` here.

File: components/convertkit/sources/new-subscriber-instant/new-subscriber-instant.mjs [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import common from "../common/common.mjs";
+
+export default {
+  ...common,
+  type: "source",
+  name: "New Subscriber Activation (Instant)",
+  key: "convertkit-new-subscriber-instant",
+  description: "Emit new event when a new subscriber is activated. [See docs here](https://developers.convertkit.com/#create-a-webhook)",
+  version: "0.0.2",
```
Comment (lcaresia) [inline components/convertkit/sources/new-subscriber-instant/new-subscriber-instant.mjs:9]:
```suggestion
  version: "0.0.1",
```

File: components/convertkit/sources/new-subscriber-instant/new-subscriber-instant.mjs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    ...common.methods,
+    getWebhookEventTypes() {
+      return {
+        "name": "subscriber.subscriber_activate",
+      };
+    },
+    async processEvent(event) {
+      const { body } = event;
+
+      if (!body?.subscriber) return;
```
Comment (lcaresia) [inline components/convertkit/sources/new-subscriber-instant/new-subscriber-instant.mjs:23]:
```suggestion
      if (!body?.subscriber) {
        return;
      }
```

File: components/convertkit/sources/new-unsubscription-instant/new-unsubscription-instant.mjs [hunk 10]
```diff
@@ -0,0 +1,33 @@
+import common from "../common/common.mjs";
+
+export default {
+  ...common,
+  type: "source",
+  name: "New Subscriber Activation (Instant)",
+  key: "convertkit-new-unsubscription-instant",
+  description: "Emit new event when a user  unsubscribers. [See docs here](https://developers.convertkit.com/#create-a-webhook)",
+  version: "0.0.2",
```
Comment (lcaresia) [inline components/convertkit/sources/new-unsubscription-instant/new-unsubscription-instant.mjs:9]:
```suggestion
  version: "0.0.1",
```

File: components/convertkit/sources/new-unsubscription-instant/new-unsubscription-instant.mjs [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    ...common.methods,
+    getWebhookEventTypes() {
+      return {
+        "name": "subscriber.subscriber_unsubscribe",
+      };
+    },
+    async processEvent(event) {
+      const { body } = event;
+
+      if (!body?.subscriber) return;
```
Comment (lcaresia) [inline components/convertkit/sources/new-unsubscription-instant/new-unsubscription-instant.mjs:23]:
```suggestion
      if (!body?.subscriber) { 
        return;
      }
```

File: components/convertkit/sources/webhook-events/webhook-events.mjs [hunk 12]
```diff
@@ -0,0 +1,104 @@
+import common from "../common/common.mjs";
+import constants from "../common/constants.mjs";
+
+export default {
+  ...common,
+  key: "convertkit-webhook-events",
+  name: "New Webhook Event (Instant)",
+  description: "Emit new event for each selected event types. [See docs here](https://developers.convertkit.com/#create-a-webhook)",
+  type: "source",
+  version: "0.0.2",
```
Comment (lcaresia) [inline components/convertkit/sources/webhook-events/webhook-events.mjs:10]:
```suggestion
  version: "0.0.1",
```

**Review by lcaresia — CHANGES_REQUESTED** [review]
Thanks for your contribution!

Just minors fixes!

**Review by lcaresia — APPROVED** [review]
LGTM!

**Comment by vunguyenhung** [comment]
Hello everyone, I have tested this PR and there're some test cases failed or needed improvement. 

Please check the test report below for more information
[Coverkit_3486_2046.pdf](https://github.com/PipedreamHQ/pipedream/files/9190341/Coverkit_3486_2046.pdf)

**Comment by vunguyenhung** [comment]
Hi everyone, all test cases are passed! Ready for release! 

Test report
[Coverkit_3486_1836.pdf](https://github.com/PipedreamHQ/pipedream/files/9198125/Coverkit_3486_1836.pdf)

**Review by jcortes — APPROVED** [review]
Approved!

**Pull request merged**